bun/src/query/queryUtils.ts

```diff
@@ -3,66 +3,70 @@ import { ValueOf } from "type-fest";
 import type { TraverserMap, GroovyTraversal } from "groovy/dsl";
 
 /**
- * Generic return type for a gremlin traversal
- * currently supports {@link GroovyTraversal} for chaining and {@link TraverserMap} for results
+ * Generic return type for a {@link GroovyTraversal} invocation
+ * currently supports {@link GroovyTraversal} for chaining and {@link TraverserMap} for traversal results
  */
 export interface NextT {
   GT: GroovyTraversal;
   TM: TraverserMap;
 }
+
 /**
  * union type of {@link NextT} object values
  */
 export type NextTUnion = ValueOf<NextT>;
 
 /**
  * base opts for a gremlin traversal
+ * @prop end if truthy returns a traversal value, else returns a traversal for chaining
  * @prop limitX e.g. traversal.range(limitX, limitY)
  * @prop limitY e.g. traversal.range(limitX, limitY)
- * @prop end if truthy returns a traversal value, else returns a traversal for chaining
  */
 export type BaseOpts<T = Record<string, any>> = T & {
+  end?: boolean;
   limitX?: number;
   limitY?: number;
-  end?: boolean;
 };
 
 /**
- * helper fn for supplying options to a gremlin traversal
+ * helper fn for supplying options to a {@link GroovyTraversal}
  * @param overrides
  * @returns
  */
 export const getBaseOpts = <T>(overrides: BaseOpts<T>) => ({
   limitX: 0,
   limitY: (overrides.limitX ?? 0) + 10,
+  end: true,
   ...overrides,
 });
 
 export type Next = { gt: NextT["GT"]; end?: BaseOpts["end"] };
+export type NextResult<T> = { value: Map<string, T>; done: boolean };
 /**
  * takes a {@link GroovyTraversal} and returns either a {@link GroovyTraversal} or {@link TraverserMap}
  * @see {@link NextT} for potential return types
  */
 export const next = <T = NextTUnion>({
   gt,
   end = true,
-}: Next): T extends NextT["GT"] ? T : Promise<T[]> => {
+}: Next): T extends NextT["GT"] ? T : Promise<NextResult<T>> => {
+  // FYI using next() > .elementMap().toList() to enable more complex chaining
   // @ts-ignore dunno how to fix this
```

Comment:
this shiz isnt scalable, typescript should automatically detect the return type
- try to overload the fn
- create distinct fns for each and return that shiz
- or just get the return type to work ;)~